pkg/deploy/generator/scripts/commonVMSS.sh

```diff
@@ -0,0 +1,1281 @@
+        )
+
+        log "Running az login with retries"
+        retry cmd retry_time
+    fi
+
+    # Suppress emulation output for podman instead of docker for az acr compatability
+    mkdir -p /etc/containers/
+    mkdir -p /root/.docker
+    touch /etc/containers/nodocker
+
+    # This name is used in the case that az acr login searches for this in it's environment
+    export REGISTRY_AUTH_FILE="/root/.docker/config.json"
+    
+    if [ -n "${registry_conf}" ]; then
+        write_file REGISTRY_AUTH_FILE registry_conf true
+    fi
+
+    log "logging into prod acr"
+    if $az_login; then
+        cmd=(
+            az
+            acr
+            login
+            --name
+            "$(sed -e 's|.*/||' <<<"$ACRRESOURCEID")"
+        )
+
+        log "Running az login with retries"
+        retry cmd retry_time
+    fi
+
+    # shellcheck disable=SC2068
+    for i in ${pull_images[@]}; do
+        local -n image="$i"
+        cmd=(
+            podman
+            pull
+            "$image"
+        )
+
+        log "Pulling image $image with retries now"
+        retry cmd retry_time
+    done
+
+    if $az_login; then
+        cmd=(
+            az
+            logout
+        )
+
+        log "Running az logout with retries"
+        retry cmd retry_time
+    fi
+}
+
+# enable_services enables all services required for aro rp
+# args:
+# 1) services - array; services to be enabled
+enable_services() {
+    local -n services="$1"
+    log "starting"
+
+    systemctl daemon-reload
+
+    log "enabling services ${services[*]}"
+    # shellcheck disable=SC2068
+    for service in ${services[@]}; do
+        log "Enabling and starting $service now"
+        systemctl enable \
+                  --now \
+                  "$service"
+    done
+}
+
+# get_reboot_offset
+# Parses the last character from the short hostname, and assigns an offset to minutes
+# Each increase in the hostname's suffix by 1 adds 5 minutes, starting with 0
+# args:
+# 1) minutes - nameref, integer; Empty variable used to hold the reboot offset minutes
+get_reboot_offset() {
+    local -n minutes="$1"
+
+    host="${HOSTNAME: -1}"
+    local -i minutes_offset=0
+    # A loop is used to acccount for scaling up the VMSS after the initial 0-2 are created, rather than a case statement
+    for i in {0..100}; do
+        if [ "$host" -eq "$i" ]; then
+            minutes="$minutes_offset"
+        fi
+        ((minutes_offset+=5))
+    done
+}
+
+# reboot_vm restores all selinux file contexts, then schedules a reboot for one hour later
+# Reboots should scheduled after all VM extensions have had time to complete
+# Reference: https://learn.microsoft.com/en-us/azure/virtual-machines/extensions/custom-script-linux#tips
+reboot_vm() {
+    log "starting"
+
+    configure_selinux "true"
+    
+    local -i offset
+    get_reboot_offset offset
+    hour="$(date -d "1 hour" +%H:%M + $offset minutes)"
+    shutdown -r "$hour" "Post deployment reboot is happening at $hour"
```

Comment:
Just a heads up - Microsoft recommends not putting reboots into CSE scripst: https://learn.microsoft.com/en-us/azure/virtual-machines/extensions/custom-script-linux

This is more of a tip than a requirement, but maybe it's something we can clean up later.

Reply:
Technically a reboot isn't happening in this script. It's being scheduled for later. The existing scripts have the reboot occurring in the script.

With the changes in this script, everything is functional in the present runtime. With the existing scripts in master, they are persisted to disk, but not started, such as enabling but not starting services, firewalld rules.

The point in rebooting now is simply to ensure a reboot doesn't impact the running state. Incase an instance were to reboot.
If we don't test that during deployment, Can we safely assume the instances will never reboot?

Reply:
@niontive I've broken `commonVMSS.sh` up into 4 different `util-*.sh` files.